memory and flash optimizations for 3.12 PR2

leptun opened the pull request:
Supersedes the broken PR https://github.com/prusa3d/Prusa-Firmware/pull/3181
Comments from there still apply.

Stats as of https://github.com/prusa3d/Prusa-Firmware/pull/3365/commits/f068c3258f477a114d4b59ad36f83f6ae4f023c7 for MK3S build:
- flash: 232522 -> 229386 = 3136B
- ram: 6115 -> 5679 = 436B

leptun committed:
Remove unused `skip_debug_msg` flag

leptun committed:
Tmc2130 use ShortTimer instead of custom timer implementation

leptun committed:
Remove unused Stepper.cpp variables

leptun committed:
TimeNow && TimeSent

leptun committed:
nIRsensorLastTime

leptun committed:
Adjust temperature runaway extruder count

leptun committed:
Change arduino IO to fastio where possible

leptun committed:
Remove maxlimit_status

leptun committed:
Fix build

leptun committed:
Disable PRUSA M28

leptun committed:
Sound enums as uint8_t

leptun committed:
Disable unimplemented PID_ADD_EXTRUSION_RATE

leptun committed:
M42 optimization

leptun committed:
Fix M226

leptun committed:
Fix build

gudnimg committed:
Use uint8_t when using axis enum

Saves 76 bytes of flash memory

gudnimg committed:
Use memset instead of nested for-loop to zero a 2d array

Saves 26 bytes of flash memory and removes two 'int' types

gudnimg committed:
Improve lcd_selftest_screen_step() parameter types

Saves 62 bytes of flash memory

gudnimg committed:
1. Remove redundant variable 'inters'
2. 'gh' variable can be 1 byte instead of 2

This saves 26 bytes of flash memory

gudnimg committed:
Remove unused global variable 'chunkHead'

No change in memory on my end, but it is one less 'int' :)

gudnimg committed:
Various optimisation to change int to uint8_t

This commit saves 174 bytes of flash memory

gudnimg committed:
Change temp_runaway_status from float to uint8_t and its enum

Saves 90 bytes of flash, and 12 bytes of SRAM

gudnimg committed:
Change temp_runaway_timer from float to uint32_t

Saves 32 bytes of flash

Also change temp_runaway_error_counter from int to
uint16_t to be specific

gudnimg committed:
Remove unused variable

Saves 2 bytes of SRAM

gudnimg committed:
Change two arrays from int to uint8_t

__preheat_counter has max value of 16
__preheat_errors has max value of 5

Saves 58 bytes of flash memory and 4 bytes of SRAM

gudnimg committed:
temp_runaway_check() parameter _heater_id should be uint8_t

Saves 58 bytes of flash

leptun committed:
Merge pull request #6 from gudnimg/gudni-PR1

Purged some `int`s for `uint8_t` when using the axis enum

leptun committed:
Fix LUT loop

gudnimg committed:
Use memset to zero previous_speed array

Saves 150 bytes of flash memory

gudnimg committed:
Use inlined function set_destination_to_current() in more places

Saves 70 bytes of flash memory

gudnimg committed:
Replace for-loop with memcpy

Saves 22 bytes of flash memory

gudnimg committed:
Change row_offsets[] to uint8_t

Saves 10 bytes of flash, and 4 bytes of SRAM

gudnimg committed:
Change return types of two functions uint8_t

These functions aren't used but they should return the appropriate type.

gudnimg committed:
Replace three lines by one function call which does the same thing

Saves 42 bytes of flash memory

gudnimg committed:
setTargetHotend expect an uint8_t, not int for second parameter

also change _usb_timer to ShortTimer

Saves 28 bytes of Flash and 1 byte of SRAM

gudnimg committed:
Convert mmu_last_request and mmu_last_response to ShortTimer

Saves 180 bytes of flash and 2 byte of SRAM

gudnimg committed:
Convert previous_millis_cmd to LongTimer

Saves 196 bytes of Flash but adds 1 byte of SRAM

gudnimg committed:
Convert autostart_atmillis to ShortTimer

Saves 34 bytes of flash and 1 byte of SRAM

gudnimg committed:
Convert NcTime to ShortTimer

The variable is only used in ultralcd.cpp so I made it static there
and deleted it from Marlin_main.cpp

Saves 24 bytes of flash and 1 byte of SRAM

gudnimg committed:
Convert mmu_last_finda_response to ShortTimer

Saves 48 bytes of Flash, 1 byte of SRAM

gudnimg committed:
Convert display_time to ShortTimer

Saves 24 bytes of flash, 1 byte of SRAM

gudnimg committed:
crashDetTimer can be ShortTimer

Expired value is less than 65535

Saves 8 bytes of flash and 2 bytes of SRAM

gudnimg committed:
Remove unused kicktime timer

The variable is only set and never actually used

Saves 8 bytes of flash and 4  bytes of SRAM

gudnimg committed:
Remove internal Prusa command 'Beat'

It doesn't seem to do anything.

Saves 22 bytes of flash

gudnimg committed:
Change usb_printing_counter to uint8_t

It is assigned max value of 10

Saves 26 byte of flash and 1 byte of SRAM

gudnimg committed:
Change heating_status to uint8_t

Saves 66 byte of flash and 1 byte of SRAM

gudnimg committed:
Change heating_status_counter to uint8_t

It has a max value of 14

Saves 28 bytes of flash and 1 byte of SRAM

gudnimg committed:
Change custom_message_state_old  and custom_message_state to uint8_t

I see max value as 7*7+10 = 59

Saves 100 byte of flash and 1 byte of SRAM

gudnimg committed:
Change farm_timer to uint8_t

Remove extern declaration of farm_timer and farm_status as they
are only used in one file.

Saves 26 bytes of flash

gudnimg committed:
Create enum class for heating_status variable

gudnimg committed:
Remove variable move_menu_scale

The variable is just set to 1.0 and is never modified, I also
removed the function lcd_move_menu_1mm()

Changes save 96 bytes of flash memory and 4 bytes of SRAM

gudnimg committed:
Smaller code in G28

Replace two blocks of code with already defined functions.

Change saves 100 bytes of flash memory and 11 lines of code.

gudnimg committed:
G81: Change for loop variables x and y to uint8_t

Saves 198 bytes of flash, that's crazy

gudnimg committed:
Use code_value_short()  in get_command()

Same code but uses a defined function

Saves 28 bytes of flash memory

gudnimg committed:
Use code_value_uint8() in get_command()

The if statement is simplified as the value is never negative
and we can get rid of one variable

Saves 12 bytes of flash memory

leptun committed:
Merge pull request #7 from gudnimg/gudni-PR2

Save some memory by using memset and inlined memcpy

gudnimg committed:
Remove one unused variable

Saves 1 byte of SRAM

The variable is only assigned a value but is never used.

gudnimg committed:
Fix an issue found in code review

Unfortunately this increases flash usages by 212 bytes (same usage as in current release) It seems it is most memory effcient to use int16_t (int). int8_t requires more memory.

gudnimg committed:
Change set_z() parameters to uin8_t

Saves 24 bytes of flash

gudnimg committed:
Remove undefined function

No change in memory footprint.

gudnimg committed:
Loop index can be unsigned

Saves 4 bytes of flash memory

leptun committed:
Fix G81 for loop

leptun committed:
Merge pull request #8 from gudnimg/remove-unused-var-gudni

Remove unused variables
Shrink integer types where possible

gudnimg committed:
Merge remote-tracking branch 'upstream/MK3' into rebase-testing-ground-v2

leptun commented:
@gudnimg Now we need to check that nothing got broken during the rebase. To begin, let's make it build again

leptun committed:
Fix temperature runaway variable types (merge conflict)

leptun committed:
Fix build (merge conflict)

gudnimg commented on Firmware/cardreader.h:
```diff
@@ -130,13 +130,12 @@ class CardReader
   
-  uint16_t nrFiles; //counter for the files in the current directory and recycled as position counter for getting the nrFiles'th name in the directory.
-  char* diveDirName;
+  int16_t nrFiles; //counter for the files in the current directory and recycled as position counter for getting the nrFiles'th name in the directory.
```
`nrFiles` should be `uint16_t`. It must have been reverted to `int16_t` 🤔

gudnimg replied:
https://github.com/prusa3d/Prusa-Firmware/pull/3245

leptun replied:
I hope other already merged PRs weren't affected during the rebase. If we see anything suspicious in this PR, we must investigate

gudnimg committed:
Fix type of nrFiles

gudnimg commented:
@leptun is the PR locked or is it Ok to add more optimisations 🙃? I may have found a way to optimise memory by removing static keywords on local function variables. Also found another int which can be uint8_t

leptun commented:
@gudnimg You can still add stuff 👍

leptun committed:
Optimize MUL8x16R8

leptun commented:
@gudnimg I managed to optimize MultiU16X8toH16 more.
With arduino 1.8.5 I optimized 14B.
With arduino 1.8.13 I optimized 32B since the compiler is not hot garbage. I've had the code increase (!!!) with GCC 5.4.0 at times when I was allowing the compiler to choose what register to use instead of R26. Now that I don't use R26, the code is somewhat more optimized even with the old compiler. The new compiler was optimizing correctly even when I still used R26, so it's just the old compiler that fails to optimize correctly.

gudnimg committed:
Optimise if-statments in world2machine_clamp

Only one of the statements can be true: tmpx < X_MIN_POS or tmpx > X_MAX_POS.
So we can be a little bit smarter here and skip the second if statement if the first was true.

This saves 6 bytes of flash memory and potential some clock cycles

gudnimg committed:
Merge branch 'MK3_3.12_Optimizations' of https://github.com/leptun/Prusa-Firmware into MK3_3.12_Optimizations

gudnimg committed:
Optimise lcd_selftest_screen input parameters and return type

* _progress_scale is max 3 so it can be uint8_t
* _progress is max 6 so it can be uint8_t
* Change _delay type to uint16_t to be explicit

Changes save 102 bytes of flash memory (Arduino IDE 1.8.19)

gudnimg committed:
Change prusa_stat_printerstatus parameter to uint8_t

Also changes status_number global variable to uint8_t

Changes save 80 bytes of flash and 1 byte of SRAM (Arduino IDE 1.8.19)

gudnimg committed:
Remove redundant variable farm_status
It is unused and status_number variable can be used instead.

leptun committed:
Remove "hh" in fmt where it makes no difference

leptun committed:
Simplify GETPC()

leptun commented:
@gudnimg Let's try and find any places in the code that could use __uint24 instead of uint32_t/long/whatever. I'm pretty sure some place in the stepper routines could use 24bit integers instead of 32bit (the acceleration variables)

gudnimg committed:
Optimise usage of code_value()

code_value() is float but in some cases we can save memory when the expected output is only 1 or 2 bytes.

Changes save 182 bytes of flash memory on my end.

gudnimg committed:
Merge branch 'MK3_3.12_Optimizations' of https://github.com/leptun/Prusa-Firmware into MK3_3.12_Optimizations

gudnimg commented:
I managed to save a good chunk of bytes by looking over how we use `code_value()`. A bit overkill converting a `float` to `uint8_t` in a few cases :P

leptun committed:
SD sorting entries instead of positions

leptun commented:
Got 200B of ram and 44B of flash by changing how the sorting works. I became aware of a limitation of fat32 which allowed me to change from uint32_t addressing to uint16_t for file entries in directories. Since 400B are thrown away for soring, 200B are saved when we halve that cache

leptun committed:
remove firstrun

leptun commented:
@gudnimg We'll probably merge this PR to MK3 on Monday. Unless you have some other changes already made, I think we should stop optimizing for now and verify the PR over the weekend.

gudnimg commented:
@leptun I have no changes locally.

leptun committed:
Pack endstop flags

leptun commented:
I forgot to push this commit. Now it should be final

vintagepc reviewed (approved):
Looks reasonable to me, I don't see any glaring issues.

leptun committed:
Fix Welcome message and language flashing

leptun committed:
Also change MUL24x24R24 to the new format and fix rounding

gudnimg commented:
This tiny change saves 64 bytes on my end <img width="552" alt="image" src="https://user-images.githubusercontent.com/8218499/152663480-0effe864-1258-493c-96a8-632653a6ee8f.png">

leptun commented:
Please add it 👍

gudnimg committed:
Optimise _menu_edit_P

If the first if statement is true, then we don't need to check the next if statement.

Saves 64 bytes of flash memory

gudnimg commented:
Added another commit for an optimization I found last night since it saves > 100 bytes. 

For reference here is the description of `hypot()`:
<img width="919" alt="image" src="https://user-images.githubusercontent.com/8218499/152675185-7fdedd53-676d-4b4a-96d8-c8704ee250bd.png">

leptun commented:
Are you 100% sure this has no side effects (like loss in precision or something like that)?
I'd like to stop with the changes for now so that we can finally test and merge this PR tomorrow. I have ~9 PRs that I want to merge tomorrow

gudnimg commented:
> Are you 100% sure this has no side effects (like loss in precision or something like that)?
> I'd like to stop with the changes for now so that we can finally test and merge this PR tomorrow. I have ~9 PRs that I want to merge tomorrow

@leptun I don’t want to promise it without testing it in more detail. But I would be surprised if it affects the precision. Feel free to revert the commit, I realise this may be a risky commit so close to the expected merge date.

leptun commented:
I'll copy this commit to another branch (as in part 3 of this PR) so that we don't lose it. Then I'll drop it

leptun commented:
See https://github.com/leptun/Prusa-Firmware/tree/MK3_3.12_Optimizations_extra

leptun committed:
Fix MUL24x24R24 comment

DRracer commented on Firmware/Marlin_main.cpp:
```diff
@@ -6120,7 +6120,7 @@ if(eSoundMode!=e_SOUND_MODE_SILENT)
         if (code_seen('P'))
           pin_number = code_value();
-        for(int8_t i = 0; i < (int8_t)(sizeof(sensitive_pins)/sizeof(int8_t)); i++)
+        for(int8_t i = 0; i < (int8_t)(sizeof(sensitive_pins)/sizeof(*sensitive_pins)); i++)
```
please use
`sizeof(sensitive_pins)/sizeof(sensitive_pins[0])`

I know it has the same effect, but its intent is much more obvious

leptun replied:
done

DRracer commented on Firmware/planner.cpp:
```diff
@@ -459,10 +459,7 @@ void plan_init() {
-  previous_speed[1] = 0.0;
-  previous_speed[2] = 0.0;
-  previous_speed[3] = 0.0;
+  memset(previous_speed, 0, sizeof(previous_speed));
```
I'm surprised the compiler couldn't optimize the float assignments into a single block of "memset"
Have you checked the assembly from this optimization? Does it call the `memset` or has it been inlined in a form of a short zeroing cycle?

leptun replied:
Before:
```
  previous_speed[0] = 0.0;
   3152a:	sts	0x0532, r1	; 0x800532 <_ZL14previous_speed.lto_priv.327>
   3152e:	sts	0x0533, r1	; 0x800533 <_ZL14previous_speed.lto_priv.327+0x1>
   31532:	sts	0x0534, r1	; 0x800534 <_ZL14previous_speed.lto_priv.327+0x2>
   31536:	sts	0x0535, r1	; 0x800535 <_ZL14previous_speed.lto_priv.327+0x3>
  previous_speed[1] = 0.0;
   3153a:	sts	0x0536, r1	; 0x800536 <_ZL14previous_speed.lto_priv.327+0x4>
   3153e:	sts	0x0537, r1	; 0x800537 <_ZL14previous_speed.lto_priv.327+0x5>
   31542:	sts	0x0538, r1	; 0x800538 <_ZL14previous_speed.lto_priv.327+0x6>
   31546:	sts	0x0539, r1	; 0x800539 <_ZL14previous_speed.lto_priv.327+0x7>
  previous_speed[2] = 0.0;
   3154a:	sts	0x053A, r1	; 0x80053a <_ZL14previous_speed.lto_priv.327+0x8>
   3154e:	sts	0x053B, r1	; 0x80053b <_ZL14previous_speed.lto_priv.327+0x9>
   31552:	sts	0x053C, r1	; 0x80053c <_ZL14previous_speed.lto_priv.327+0xa>
   31556:	sts	0x053D, r1	; 0x80053d <_ZL14previous_speed.lto_priv.327+0xb>
  previous_speed[3] = 0.0;
   3155a:	sts	0x053E, r1	; 0x80053e <_ZL14previous_speed.lto_priv.327+0xc>
   3155e:	sts	0x053F, r1	; 0x80053f <_ZL14previous_speed.lto_priv.327+0xd>
   31562:	sts	0x0540, r1	; 0x800540 <_ZL14previous_speed.lto_priv.327+0xe>
   31566:	sts	0x0541, r1	; 0x800541 <_ZL14previous_speed.lto_priv.327+0xf>
```
After:
```
  memset(previous_speed, 0, sizeof(previous_speed));
   3152c:	ldi	r30, 0x32	; 50
   3152e:	ldi	r31, 0x05	; 5
   31530:	movw	r26, r30
   31532:	st	X+, r1
   31534:	dec	r24
   31536:	brne	.-6      	; 0x31532 <main+0xae0>
```

DRracer commented on Firmware/Marlin_main.cpp:
```diff
@@ -2827,7 +2827,7 @@ static void gcode_G28(bool home_x_axis, long home_x_value, bool home_y_axis, lon
       enable_endstops(true);
 
-      memcpy(destination, current_position, sizeof(destination));
+      set_destination_to_current();
```
:+1:

DRracer commented on Firmware/Marlin.h:
```diff
@@ -275,6 +275,17 @@ FORCE_INLINE unsigned long millis_nc() {
 void setPwmFrequency(uint8_t pin, int val);
 #endif
 
+enum class HeatingStatus : uint8_t
```
:+1:

DRracer commented on Firmware/Marlin_main.cpp:
```diff
@@ -5623,7 +5623,7 @@ if(eSoundMode!=e_SOUND_MODE_SILENT)
                 SERIAL_PROTOCOL(MESH_HOME_Z_SEARCH);
                 SERIAL_PROTOCOLLNPGM("\nMeasured points:");
-                for (uint8_t y = MESH_NUM_Y_POINTS-1; y >= 0; y--) {
+                for (int16_t y = MESH_NUM_Y_POINTS-1; y >= 0; y--) {
```
this looks like something else has not been optimized - it doesn't make sense to keep the int16_t at this spot if `MESH_NUM_Y_POINTS` never exceeds 255.

leptun replied:
this was fixed in a later commit:
```
                for (uint8_t y = MESH_NUM_Y_POINTS; y-- > 0;) {
                    for (uint8_t x = 0; x < MESH_NUM_X_POINTS; x++) {
```

DRracer commented on Firmware/mesh_bed_calibration.cpp:
```diff
@@ -2872,14 +2872,14 @@ bool sample_mesh_and_store_reference()
         mbl.set_z(0, 0, current_position[Z_AXIS]);
     }
-    for (int8_t mesh_point = 1; mesh_point != MESH_MEAS_NUM_X_POINTS * MESH_MEAS_NUM_Y_POINTS; ++ mesh_point) {
+    for (uint8_t mesh_point = 1; mesh_point != MESH_MEAS_NUM_X_POINTS * MESH_MEAS_NUM_Y_POINTS; ++ mesh_point) {
```
please add a `static_assert(MESH_MEAS_NUM_X_POINTS * MESH_MEAS_NUM_Y_POINTS < 255, "overflow.....")`

leptun replied:
done

DRracer replied:
I thought you'd add a reasonable message as well why the static assert fails :wink: 
Something like:
_MESH_MEAS_NUM_X_POINTS * MESH_MEAS_NUM_Y_POINTS > 255: it would overflow the uint8_t range - the following for-cycle would misbehave_

DRracer commented on Firmware/Marlin_main.cpp:
```diff
@@ -11130,7 +11130,7 @@ void uvlo_()
 
     // Store the mesh bed leveling offsets. This is 2*7*7=98 bytes, which takes 98*3.4us=333us in worst case.
-    for (int8_t mesh_point = 0; mesh_point < MESH_NUM_X_POINTS * MESH_NUM_Y_POINTS; ++ mesh_point) {
+    for (uint8_t mesh_point = 0; mesh_point < MESH_NUM_X_POINTS * MESH_NUM_Y_POINTS; ++ mesh_point) {
```
again - please add a static_assert that MESH_NUM_X_POINTS * MESH_NUM_Y_POINTS never exceeds 255

leptun replied:
One should be enough

DRracer commented on Firmware/cardreader.h:
```diff
@@ -135,7 +135,7 @@ class CardReader
   bool autostart_stilltocheck; //the sd start is delayed, because otherwise the serial cannot answer fast enought to make contact with the hostsoftware.
   
-  int16_t nrFiles; //counter for the files in the current directory and recycled as position counter for getting the nrFiles'th name in the directory.
+  uint16_t nrFiles; //counter for the files in the current directory and recycled as position counter for getting the nrFiles'th name in the directory.
```
are you absolutely sure that `nrFiles` is never abused with a negative value?

leptun replied:
Just looked again. Nothing uses it as a signed number. Only as a file count / file index (0 indexed)

DRracer commented on Firmware/speed_lookuptable.h:
```diff
@@ -8,30 +8,36 @@ extern const uint16_t speed_lookuptable_slow[256][2] PROGMEM;
+    "adc %B0, r1" "\n\t"    //propagate carry of addition (add 0 with carry)
+    : "=&r" (out)
+    : "r" (x), "r" (y)
+    : "r0", "r1"            //clobbers: Technically these are either scratch registers or always 0 registers, but I'm making sure the compiler knows just in case.
```
nice :+1: ... looks correct to me

DRracer commented on Firmware/ConfigurationStore.cpp:
```diff
@@ -32,7 +32,7 @@ static bool EEPROM_writeData(uint8_t* pos, uint8_t* value, uint8_t size, const c
 {
 #ifdef DEBUG_EEPROM_WRITE
-	printf_P(PSTR("EEPROM_WRITE_VAR addr=0x%04x size=0x%02hhx name=%s\n"), pos, size, name);
+	printf_P(PSTR("EEPROM_WRITE_VAR addr=0x%04x size=0x%02x name=%s\n"), pos, size, name);
```
please use formatting flags defines from `inttypes.h` and be _very specific_ about the data type of the attributes
https://manned.org/PRIu8

leptun replied:
In variadic functions, uint8_t always gets promoted to 16bit. hh/h on avr doesn’t do anything for printf. It only does stuff for scanf.

leptun replied:
Reference: https://www.nongnu.org/avr-libc/user-manual/group__avr__stdio.html#:~:text=The%20hh%20length%20modifier%20is%20ignored%20(char%20argument%20is%20promouted%20to%20int).%20More%20exactly%2C%20this%20realization%20does%20not%20check%20the%20number%20of%20h%20symbols.

DRracer commented on Firmware/Marlin_main.cpp:
```diff
@@ -6511,9 +6511,9 @@ SERIAL_PROTOCOLPGM("\n\n");
     {
-        if(code_seen('P')) print_percent_done_normal = code_value();
+        if(code_seen('P')) print_percent_done_normal = code_value_uint8();
         if(code_seen('R')) print_time_remaining_normal = code_value();
```
please rename `code_value` to `code_value_float` so that we don't run into such an issue in the future.

leptun replied:
We already have code_value_float with a workaround for E notation. How should we proceed?

DRracer commented on Firmware/cardreader.cpp:
```diff
@@ -689,11 +689,11 @@ void CardReader::getfilename(uint16_t nr, const char * const match/*=NULL*/)
 	curDir = &workDir;
 	nrFiles = 0;
-	curDir->seekSet(position);
+	curDir->seekSet((uint32_t)entry << 5);
```
nice :+1:

DRracer commented on Firmware/eeprom.cpp:
```diff
@@ -100,6 +100,11 @@ if (eeprom_read_byte((uint8_t*)EEPROM_PINDA_TEMP_COMPENSATION) == 0xff) eeprom_u
 	if (eeprom_read_dword((uint32_t*)EEPROM_JOB_ID) == EEPROM_EMPTY_VALUE32)
 		eeprom_update_dword((uint32_t*)EEPROM_JOB_ID, 0);
+
+    if (eeprom_read_byte((uint8_t *)EEPROM_TOTALTIME) == 255 && eeprom_read_byte((uint8_t *)EEPROM_TOTALTIME + 1) == 255 && eeprom_read_byte((uint8_t *)EEPROM_TOTALTIME + 2) == 255 && eeprom_read_byte((uint8_t *)EEPROM_TOTALTIME + 3) == 255) {
```
Wasn't this the way to "create" negative filament usage in stats? In case the user skipped the wizard or something?

leptun replied:
I don't understand. I simply moved the initialization of a eeprom variable from the status screen to eeprom_init(). I don't see the issue in doing this.

DRracer commented on Firmware/ultralcd.cpp:
```diff
@@ -772,21 +771,6 @@ void lcdui_print_status_screen(void)
 // Main status screen. It's up to the implementation specific part to show what is needed. As this is very display dependent
 void lcd_status_screen()                          // NOT static due to using inside "Marlin_main" module ("manage_inactivity()")
 {
-	if (firstrun == 1) 
```
I'm not completely convinced that the removal of this code is correct. Please prepare a detailed feature test for this scenario.

leptun replied:
I'll do a factory reset later and look at the statistics. i already tested the welcome message change, so that is fine, but haven't tested the statistics init yet. Even though that should be fine, I'll test it anyway a bit later

DRracer commented on Firmware/stepper.cpp:
```diff
@@ -134,9 +129,7 @@ static uint16_t OCR1A_nominal;
-static volatile bool endstop_x_hit=false;
-static volatile bool endstop_y_hit=false;
-static volatile bool endstop_z_hit=false;
+static volatile uint8_t endstop_hit = 0;
```
Please DON'T - these flags are used in time critical code and updating a single bit takes more CPU cycles than writing a single byte.

leptun replied:
I disagree. Operations on multiple bits at a time are much faster.

This is quicker:
```
endstop_hit = 0;
   3262e:	sts	0x0574, r1	; 0x800574 <_ZL11endstop_hit.lto_priv.323>
###################
 if( endstop_hit) {
   3253a:	lds	r24, 0x0574	; 0x800574 <_ZL11endstop_hit.lto_priv.323>
   3253e:	and	r24, r24
```
Single bit read is Just as fast as before because it had to turn a byte into a bool. Now it uses sbrs instead of tst:
```
   if(endstop_hit & _BV(X_AXIS)) {
   32554:	lds	r24, 0x0574	; 0x800574 <_ZL11endstop_hit.lto_priv.323>
   32558:	sbrs	r24, 0
```

Single bit changes are indeed slower by default. I see where it could be a problem in stepper_check_endstops(), but I just managed to save 192B of flash just by caching volatile values in registers while inside stepper_check_endstops (perfectly safe because of the interrupt part). It's in fact now a lot faster than before. even with caching since I read and write to memory only once and otherwise just do ORI and ANDI on a register

Tested the functionality again using Home, Belt test, Z calibration and MBL. All worked correctly.

DRracer commented on Firmware/speed_lookuptable.h:
```diff
@@ -39,61 +39,70 @@ FORCE_INLINE uint16_t MUL8x16R8(uint8_t x, uint16_t y) {
+    "add r27, r1 \n\t"
+    "adc %A0, r26 \n\t"
+    "adc %B0, r26 \n\t"
+    "lsl r27 \n\t"
```
please explain the change from `lsr` to `lsl`

leptun replied:
@DRracer This is my conversation with Yuri about this

--leptun
Hello :wave:
I was looking at the speed lookup table asm functions and noticed some weird behavior:
```
	MultiU16X8toH16(x, 100, 42); // 0x1068 >>8 => 0x10
	MultiU16X8toH16(x, 101, 42); // 0x1092 >>8 => 0x10
	MultiU16X8toH16(x, 101, 43); // 0x10f7 >>8 => 0x11 ???
	MultiU16X8toH16(x, 95, 41);  // 0x0f37 >>8 => 0x10 ???
	MultiU16X8toH16(x, 95, 42);  // 0x0f96 >>8 => 0x0f ???
```
Why are odd numbers rounded up and even numbers rounded down? That really makes no sense. Is the function trying to round the >>8 but using the lsb instead of the msb by mistake?

--Yuri D'Elia
Huh, good question. No idea, we really need to check the asm. This sounds almost intentional :thinking_face:
Were you trying to use them for some other multiplication?

--leptun
No. I noticed I could optimize it a bit more
And while doing that I also noticed this odd behaviour
Considering this function is used for stepper timer calculation, it made me wonder if this weird rounding is introducing artifacts in accelerations

--Yuri D'Elia
These come from marlin 1.x (even before the MK2 fork)

--leptun
So they might be wrong

Yuri D'Elia
It could be. But at some point when I added the "c" code I did some for loops comparing the outputs and I didn't notice this odd rounding
I remember definitely the comment above the function was wrong. And I only understood what was it attempted to do by cloning marlin and trying to look at the history

--leptun
the c code does absolutely no rounding. So that's what's even weirder
It does a multiply and then right shifts the result. But it also fiddles with this least significant bit for some reason
Btw, here is my proposed optimization:
![image](https://user-images.githubusercontent.com/17808203/153364971-cdb3a020-06cf-4dc0-b289-e90ba75a0eae.png)
Do you see anything that could go wrong by doing so?
The old code was adding r1 to the result and then adding that rounding as the carry of "lsr". r26 was 0, so they were just adding the carry to the result
My optimization simply adds the carry at the same time r1 is added to the result, avoiding those "adc r26" commands
Both my code and the old code exhibit this weird rounding issue
If instead of "lsr r0" I do "lsl r0", then that will get the most significant bit of r0, so it actually rounds the output correctly
lsl and lsr are logical shift btw. They push either the msb or the lsb to the carry and shift the register by 1.
![image](https://user-images.githubusercontent.com/17808203/153365208-82f22f83-cdd3-4a63-a53d-2132060832db.png)
![image](https://user-images.githubusercontent.com/17808203/153365221-fbd1dfe4-4137-4f4e-b21b-cea3edffeab4.png)

Comparison:
```
	asm("nop\n\t");//                             lsr       lsl
	MultiU16X8toH16(x, 100, 42); // 0x1068 >>8 => 0x10      0x10
	MultiU16X8toH16(x, 101, 42); // 0x1092 >>8 => 0x10      0x11
	MultiU16X8toH16(x, 101, 43); // 0x10f7 >>8 => 0x11 ???  0x11
	MultiU16X8toH16(x, 95, 41);  // 0x0f37 >>8 => 0x10 ???  0x0f
	MultiU16X8toH16(x, 95, 42);  // 0x0f96 >>8 => 0x0f ???  0x10
```
We could also technically disable the rounding completely and do just like the C code, but I think doing the rounding correctly might help, unlike the broken rounding which could oscillate between two values with the right conditions

--leptun
Look how it oscillates in this scenario (with lsr like before)
![image](https://user-images.githubusercontent.com/17808203/153365431-0d70b951-41d4-4589-aa1b-2b00e238e942.png)
And this is the same, but with lsl instead of lsr
![image](https://user-images.githubusercontent.com/17808203/153365519-3e0f6418-1ea0-4794-b7a4-ac6dbdaf5b55.png)

MultiU24X24toH16 also oscillates
![image](https://user-images.githubusercontent.com/17808203/153365740-f3f31e5a-4659-46c5-9305-11a181682e40.png)
And, not surprising at all, changing the lsr there to lsl also fixes those oscillations:
![image](https://user-images.githubusercontent.com/17808203/153365785-a3e74c1a-f3f9-49bb-8edb-1388040ce8a4.png)

--leptun
I did two test prints, the surface looks the same :confused:

leptun replied:
Please keep in mind that I managed to optimize the 16x8 function even more in the meantime, but the stuff mentioned above still holds

DRracer replied:
ok, good enough, thank you

DRracer reviewed (changes requested):
I went through each commit separately. Most of them ale completely correct and contribute to lowering the code size greatly.
But I found some changes which I don't find completely correct. At least they need explanation and/or fixing.

leptun committed:
Fix sizeof array look

leptun committed:
Add static assert

leptun committed:
Optimize endstop flags even more

leptun committed:
Merge branch 'MK3' into MK3_3.12_Optimizations

leptun commented:
Also fixed merge conflict

DRracer commented on Firmware/stepper.cpp:
```diff
@@ -230,7 +230,9 @@ bool endstops_hit_on_purpose()
 bool endstop_z_hit_on_purpose()
 {
   bool hit = endstop_hit & _BV(Z_AXIS);
+CRITICAL_SECTION_START;
```
Nope, I'm strictly against it
The endstop_hit* flags must remain as standalone variables due to being used in time critical code. Any additional instructions are to be avoided.
Have a look at the assembly output if you don't believe :wink:

DRracer replied:
explained, closing this comment

leptun committed:
Faster code when endstops/probe are not enabled

leptun committed:
VERBOSE_CHECK_HIT_ENDSTOPS

leptun commented:
@gudnimg Well done with this PR 🎉 Now the intense testing shall begin with what we have so far in 3.12.0 https://github.com/prusa3d/Prusa-Firmware/milestone/1?closed=1
